### PR TITLE
AO3-6312 Add preference to prevent guest replies to comments

### PR DESCRIPTION
**Pull request by Bilka2**
# Pull Request Checklist

* [X] Have you read ["How to write the perfect pull request"](https://github.blog/2015-01-21-how-to-write-the-perfect-pull-request/)?
* [X] Have you read the [contributing guidelines](https://github.com/otwcode/otwarchive/blob/master/CONTRIBUTING.md)?
* [X] Have you added [tests for any changed functionality](https://github.com/otwcode/otwarchive/wiki/Automated-Testing)?
* [X] Have you added the [Jira](https://otwarchive.atlassian.net) issue number
  as the *first* thing in your pull request title (e.g. `AO3-1234 Fix thing`)
* [X] Do you have fewer than 5 pull requests already open? If not, please wait
  until they are reviewed and merged before creating new pull requests.

## Issue

https://otwarchive.atlassian.net/browse/AO3-6312

## Purpose

Add a preference to prevent logged out users from replying to your comments on news posts and other users' works. Guests can still reply to you on works you created or co-created.

## Testing Instructions

See Jira.

## References

Partial implementation in #4243.
Migration with departure gem based on #4528.
Help text for the admin emails feature removed because of [AO3-5043](https://otwarchive.atlassian.net/browse/AO3-5043).

## Credit

Bilka (he/him)

[AO3-5043]: https://otwarchive.atlassian.net/browse/AO3-5043?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: app/helpers/comments_helper.rb [hunk 1]
```diff
@@ -108,7 +108,8 @@ def can_reply_to_comment?(comment)
       comment_parent_hidden?(comment) ||
       blocked_by_comment?(comment) ||
       blocked_by?(comment.ultimate_parent) ||
-      guest? && admin_settings.guest_comments_off?)
+      (guest? && admin_settings.guest_comments_off?) ||
```
Comment (weeklies) [inline app/helpers/comments_helper.rb:111]: Since the cases are getting lengthy, should we refactor this to use guard clauses?

Comment (EchoEkhi) [reply app/helpers/comments_helper.rb:111]: I don't think so, it‘s still very readable to me.

**Comment by sarken** [comment]
Just a quick comment! I think this is missing a validation in the comment model.

File: spec/models/comment_spec.rb [hunk 2]
```diff
@@ -14,7 +14,7 @@
 
     it "should be invalid if exactly duplicated" do
       expect(second_comment.valid?).to be_falsy
-      expect(second_comment.errors.keys).to include(:comment_content)
+      expect(second_comment.errors.attribute_names).to include(:comment_content)
```
Comment (Bilka2) [inline spec/models/comment_spec.rb:17]: [ActiveModel::Errors::keys](https://api.rubyonrails.org/v6.1.7/classes/ActiveModel/Errors.html#method-i-keys) is deprecated, to be removed in Rails 7. I used the opportunity to switch to the suggested replacement.

File: app/models/user.rb [hunk 3]
```diff
@@ -431,6 +431,12 @@ def no_resets?
     has_role?(:no_resets)
   end
 
+  # Whether logged out users can reply to this user's comments with the given
+  # ultimate parent
+  def disallow_guest_replies?(ultimate_parent)
```
Comment (sarken) [inline app/models/user.rb:436]: This works, but I'm not sure if it's the right approach. (Others might want to jump in and tell me I'm totally off base here.)

It's true that the user, not the comment, is the thing that controls reply permissions, so I can definitely see the logic behind putting this check in the User model.

However, we're only going to use this when we already have a comment handy, so it feels overly complicated to always need to do something like `comment.user.disallow_guest_replies?(comment.ultimate_parent)` instead of something like `comment.guest_replies_disallowed?` imo, the way we handled the block helpers (`blocked_by_comment?(comment)` and `blocked_by?(comment)`) sets a precedent for using the comment rather than the user in cases where it might be more convenient.

I think there's also some potential for confusion with this approach -- the code comment helps, but when I first saw the method, it was in the comment helper, and I thought it was checking if the user had the work (ultimate parent) set to disallow guest comments.

File: app/helpers/comments_helper.rb (hunk 1 above)
Comment (zz9pzza) [reply app/helpers/comments_helper.rb:111]: I agree with weeklies, can we refactor

File: app/models/user.rb (hunk 3 above)
Comment (zz9pzza) [reply app/models/user.rb:436]: I think I am in agreement with sarken here.

File: public/help/comment-preferences.html [hunk 4]
```diff
@@ -9,6 +9,6 @@ <h4>Comment Preferences</h4>
   <dd>Enable this option if you prefer not to receive emails for your own comments (for example, when you reply to comments on your own works).</dd>
   <dt>Turn off emails about kudos</dt>
   <dd>Enable this option if you would prefer not to receive email notifications when someone leaves kudos on your work.</dd>
-  <dt>Turn off admin emails</dt>
```
Comment (zz9pzza) [inline public/help/comment-preferences.html:12]: Why is this deleted ?

File: spec/controllers/comments_controller_spec.rb [hunk 5]
```diff
@@ -854,6 +1079,25 @@
...
+          context "when comment is a guest reply to user who turns off guest replies afterwards" do
+            let(:reply) do
+              reply = create(:comment, :by_guest, commentable: comment)
+              comment.user.preference.update!(guest_replies_off: true)
+              reply
+            end
+
+            it "freezes reply and redirects with success message" do
+              fake_login_admin(admin)
+              put :freeze, params: { id: reply.id }
```
Comment (zz9pzza) [inline spec/controllers/comments_controller_spec.rb:1092]: Where is this implemented ?

**Review by zz9pzza — CHANGES_REQUESTED** [review]

File: public/help/comment-preferences.html (hunk 4 above)
Comment (Bilka2) [reply public/help/comment-preferences.html:12]: The preference was removed 6 years ago in https://otwarchive.atlassian.net/browse/AO3-5043.

File: spec/controllers/comments_controller_spec.rb (hunk 5 above)
Comment (Bilka2) [reply spec/controllers/comments_controller_spec.rb:1092]: It's checking that the validation on the model only runs on :create. If it always runs, freezing the guest reply after the user turns off guest replies will fail validation.

I could not figure out how do this as a model test, would appreciate pointers there if that is preferred. I was hung up on how to do the freezing. Just updating the `iced` attribute didn't seem correct.

File: app/models/user.rb (hunk 3 above)
Comment (Bilka2) [reply app/models/user.rb]: I was debating where to put it, I initially decided User because it's a user preference. Anyway, happy to move it, done.

Now the comment model has two methods with very similar names, one of them used only for validation. I'm not happy with the method names, they don't feel self-explanatory for what the difference is. I could switch back to using validate_each so the validation uses a block instead of a method. Or the validation method could be private, or they could have better names. Suggestions appreciated.

File: spec/controllers/comments_controller_spec.rb (hunk 5 above)
Comment (zz9pzza) [reply spec/controllers/comments_controller_spec.rb:1092]: I am trying to understand the test and reconcile it with the code I can see.

I think the test says a comment which is from a guest becomes a frozen thread if the user of a parent comment disables commenting by guests. 

I don't think that is intended so I suspect I am missing something.

Comment (Bilka2) [reply spec/controllers/comments_controller_spec.rb:1092]: User makes comment. Guest replies. User turns off "allow guest replies". Admin tries to freeze guest reply.

Intended: Guest reply is frozen.
Possible problem this test would catch: Freezing fails because the validation of guest reply says that guest replies aren't allowed for the user.

File: app/models/comment.rb [hunk 6]
```diff
@@ -24,6 +24,19 @@ class Comment < ApplicationRecord
 
   delegate :user, to: :pseud, allow_nil: true
 
+  validate :can_be_guest_reply, if: :reply_comment?, unless: :pseud_id, on: :create
+  # Whether the writer of the comment this is replying to allows guest replies
+  def can_be_guest_reply
```
Comment (sarken) [inline app/models/comment.rb:29]: For consistency, can you put the comment above the `validate` line, like the `not_blocked` validations do?

I don't think it's super important for the methods to have very distinct names, since they're pretty closely connected -- in fact, we have both a `check_for_spam?` method and a `check_for_spam` validation in this model. But if we wanted to tweak the name, `guest_can_reply` would be in line with the `guest_cannot_kudos_restricted_work` kudos validation.

File: spec/controllers/comments_controller_spec.rb [hunk 7]
```diff
@@ -346,6 +429,74 @@
...
+        it_behaves_like "guest cannot reply to a user with guest replies disabled"
+      end
+
+      context "when comment is on user's work" do
+        let(:work) { create(:work, authors: [user.default_pseud]) }
+        let(:comment) { create(:comment, pseud: user.default_pseud, commentable: work.first_chapter) }
+
+        it_behaves_like "guest can reply to a user with guest replies disabled on user's work"
+      end
+      context "when commentable is user's co-creation" do
```
Comment (sarken) [inline spec/controllers/comments_controller_spec.rb:493]: Should there be a blank line here?

File: spec/controllers/comments_controller_spec.rb (hunk 5 above)
Comment (sarken) [reply spec/controllers/comments_controller_spec.rb:1092]: @zz9pzza It's basically trying to make sure we don't run into a problem like [AO3-6373](https://otwarchive.atlassian.net/browse/AO3-6373).

File: spec/models/comment_spec.rb [hunk 8]
```diff
@@ -167,6 +167,92 @@
...
+                  comment_content: "Hmm.")
+    end
+
+    shared_examples "creating guest reply is allowed" do
+      describe "save" do
+        it "allows guest replies" do
+          expect(guest_reply.save).to be_truthy
+          expect(guest_reply.errors.full_messages).to be_blank
+        end
+        it "allows user replies" do
```
Comment (sarken) [inline spec/models/comment_spec.rb:197]: I think we need a blank line between `end` and the next example -- same on line 210.

**Review by sarken — APPROVED** [review]
Yay, thank you! I have two spacing nitpicks but if james_ feels his comments were addressed and wants to approve and merge, that's fine, they're not vital.

**Review by zz9pzza — APPROVED** [review]
